Refresh Tachibot design

Soitora opened the pull request:
This should be prefaced with that I don't know Python at all so I just experimented until it worked and looked decent.

Estimated required bandwidth for the dominant color system = **300 MB**/month for **1000** messages (last period in Tachiyomi server).

On Windows machines, [`fast_colorthief`](https://github.com/bedapisl/fast-colorthief) has to be built manually with CMake.

## Changes
- Adds AniList embed thumbnail as the image
- Gets the image dominant color and uses it for embed color
- Changes casing on manga format
- Adds startedDate as timestamp
- Fix some quotation marks so it's more consistent
- Since `replace` is case sensitive, it wouldn't catch `<bR>`, `<Br>`, `<BR>`, `<I>` and `</I>` from descriptions – that's fixed.

## Preview
Old | New
------------ | -------------
![image](https://user-images.githubusercontent.com/10836780/105189307-1fcf3680-5b35-11eb-97f1-e813f36e149b.png) | ![image](https://user-images.githubusercontent.com/10836780/105189208-03cb9500-5b35-11eb-95a8-dd1a6a5c07ab.png)

Soitora committed:
Refresh Tachibot output design

Soitora committed:
Unify quotation marks

Soitora committed:
Fix edge cases where replace didn't catch

Case sensitive, I fixed it by adding more catches instead of trying to find a method of making them case insensitive.

Soitora committed:
Add startDate as timestamps

Also change capitalization on format

doron-goldstein commented on bot.py:
```diff
@@ -25,18 +29,34 @@ async def format_embed(self, name, anime=False):
-        desc = desc.replace("<br>", "").replace("<i>", "").replace("</i>", "")
-        footer = re.sub(r'.*\.', '', str(media.format))
-        e = discord.Embed(title=title, description=desc, color=0x4286f4)
+        desc = desc.replace("<br>", "").replace("<bR>", "").replace("<Br>", "").replace("<BR>", "").replace("<i>", "").replace("</i>", "").replace("<I>", "").replace("</I>", "")
```
This is pretty ugly. At this point, you should use regex to substitute the values. (Optimally I'd like the HTML tags to be removed anyway, but that's a little more work, and beyond this PR scope.
If you want to be extra fancy, you can replace those <br> tags with a newline (`\n`).

EDIT: A similar use case example of regex can be found on the footer line right below this one.
regex would be something along the lines of `<[bri]{1,2}>`, with the global, multiline, and case-insensitive flags.

doron-goldstein commented on bot.py:
```diff
@@ -25,18 +29,34 @@ async def format_embed(self, name, anime=False):
-        e = discord.Embed(title=title, description=desc, color=0x4286f4)
+        desc = desc.replace("<br>", "").replace("<bR>", "").replace("<Br>", "").replace("<BR>", "").replace("<i>", "").replace("</i>", "").replace("<I>", "").replace("</I>", "")
+        footer = re.sub(r".*\.", "", str(media.format))
+        title_cover_info = "https://img.anili.st/media/" + str(media.id) 
```
Should move this value into the class as a property to clean it up (and allow reuse)

doron-goldstein commented on bot.py:
```diff
@@ -25,18 +29,34 @@ async def format_embed(self, name, anime=False):
+        footer = re.sub(r".*\.", "", str(media.format))
+        title_cover_info = "https://img.anili.st/media/" + str(media.id) 
+
+        def title_cover_info_color(title_cover_info=title_cover_info,temp_file="tmp.jpg"):
```
I personally dislike having compound functions like this. Should take it out, and considering how you use every parameter as a default value, its better to actually insert them as parameters when you call the function.

doron-goldstein commented on bot.py:
```diff
@@ -25,18 +29,34 @@ async def format_embed(self, name, anime=False):
+
+        def title_cover_info_color(title_cover_info=title_cover_info,temp_file="tmp.jpg"):
+
+            opener = urllib.request.build_opener()
```
urllib is not a good choice here for 2 reasons:
- It has a very barebones and sluggish API, which makes for ugly code and annoying boilerplate. A better use would be the `requests` library, which is much more modern and is widely used as the main HTTP library in python. Though, because of the second point it won't fit us here.
- It doesn't support asynchronous operation. This means every time it has to request the image, the whole bot will wait for it to finish and only then continue listening for messages to parse. This gets worse and worse the bigger the bot/server gets, and in particularly bad cases can straight up crash the bot.

consider using an async requests library with a modern API. Kadal, the library this bot uses to query AL (which I wrote) uses either `asks` or `aiohttp` depending on the environment. Consider looking into those.

doron-goldstein commented on bot.py:
```diff
@@ -25,18 +29,34 @@ async def format_embed(self, name, anime=False):
+            urllib.request.install_opener(opener)
+            urllib.request.urlretrieve(title_cover_info, temp_file)
+
+            palette_color = fast_colorthief.get_palette(temp_file, color_count=2, quality=100)
```
This looks like it would cause the same problem `urllib` does. A synchronous operation can stall the bot. Try looking into one of two solutions.
- Find an asynchronous package that might work well here
- Run this package through an executor. This option is much worse as it reduces performance considerably.

Additionally, this library seems like a scarcely used fork of an even more dead library. I'm not really against using small time libraries but would be very much happier with a more popular one. Most of these libraries are based on PIL which in itself is synchronous, but is a very good library in general. Feel free to consult me if you're getting stuck here :)

doron-goldstein commented on bot.py:
```diff
@@ -25,18 +29,34 @@ async def format_embed(self, name, anime=False):
+            urllib.request.urlretrieve(title_cover_info, temp_file)
+
+            palette_color = fast_colorthief.get_palette(temp_file, color_count=2, quality=100)
+            embed_color = "%02x%02x%02x" % palette_color[1]
```
Use f-strings, we are running python 3 😄

doron-goldstein commented on bot.py:
```diff
@@ -25,18 +29,34 @@ async def format_embed(self, name, anime=False):
+
+            palette_color = fast_colorthief.get_palette(temp_file, color_count=2, quality=100)
+            embed_color = "%02x%02x%02x" % palette_color[1]
+            os.remove(temp_file)
```
This, again, will block the running thread. I think a library like `aiofiles` would do the trick.

Soitora committed:
Ignore src folder

Soitora committed:
Remove duplicate footer line

Soitora committed:
Use regex replace for description

Soitora committed:
Make the temp file a .tmp file

Soitora committed:
Use aiohttp/files for operations

It works, but I have no idea what I'm doing

Soitora committed:
Update script start print

Soitora committed:
Run colorthief in executor

Or something, I don't know

Co-Authored-By: de-odex <de-odex@users.noreply.github.com>

Soitora committed:
Dup id snuck in somehow

Begone

Soitora committed:
Sett anilist cover url as property

Soitora committed:
Don't italize genres

doron-goldstein commented on .gitignore:
```diff
@@ -2,4 +2,5 @@
-.vscode/
\ No newline at end of file
+.vscode/
+src/
```
What is this for?

Soitora replied:
When installing the requirements.txt on Windows this is created, and since it's a local file it shouldn't be included in the repo

doron-goldstein commented on bot.py:
```diff
@@ -25,26 +34,39 @@ async def format_embed(self, name, anime=False):
-        title = (media.title.get('english')
-                 or media.title.get('romaji')
-                 or media.title.get('native'))
-        desc = "***" + ", ".join(media.genres) + "***\n"
```
I'd like to keep the italics, it distinguishes the genres better.

Soitora replied:
I've had people complain about it (although can be a case of vocal minority). I'll run a poll on it on the server
(10 minute poll, keeping the current method won (non italics))

doron-goldstein commented on bot.py:
```diff
@@ -15,6 +22,8 @@ def __init__(self):
         self.tachi_id = 349436576037732353
         self.klient = kadal.Klient(loop=self.loop)
+        self.anilist_cover_url = "https://img.anili.st/media/"
+        print("Starting...")
```
We already have an on_ready method, this is both misplaced and misleading, as it will be called before the bot starts up.

Soitora replied:
I assumed as such (the misplacement), but right now when you press enter for the python command to start the bot it newlines and gives no indication at all that the bot is working until it has completed loading and is ready to use.

doron-goldstein commented on bot.py:
```diff
@@ -25,26 +34,39 @@ async def format_embed(self, name, anime=False):
+        desc = re.sub(r"<[bri/]{1,2}>", "", desc, flags=re.I|re.M)
+        footer = re.sub(r".*\.", "", str(media.format))
+        
+        async with aiohttp.ClientSession() as session:
```
This is my bad for noticing just now, but Kadal exposes its aiohttp/asks session so you dont have to make a new one. Just use it from `self.klient.session` ;)

doron-goldstein commented on bot.py:
```diff
@@ -25,26 +34,39 @@ async def format_embed(self, name, anime=False):
+        
+        async with aiohttp.ClientSession() as session:
+            async with session.get(f"{self.anilist_cover_url}{media.id}") as resp:
+                if resp.status == 200:
```
You need to handle the case where its not 200, this will throw, since the color is defined inside but used outside. (the image in set_image would also be invalid, but im pretty sure discord handles that on itself so its up to you on that front)

Soitora replied:
If the image doesn't load I'm pretty sure it just doesn't load at all, but we could maybe do some tests to make sure it wouldn't become an issue

doron-goldstein commented on bot.py:
```diff
@@ -25,26 +34,39 @@ async def format_embed(self, name, anime=False):
+        async with aiohttp.ClientSession() as session:
+            async with session.get(f"{self.anilist_cover_url}{media.id}") as resp:
+                if resp.status == 200:
+                    f = await aiofiles.open("image.tmp", mode='wb')
```
This can also be an `async with` context manager to avoid having to explicitly close the file (and it looks nicer)

doron-goldstein commented on bot.py:
```diff
@@ -25,26 +34,39 @@ async def format_embed(self, name, anime=False):
+                    f = await aiofiles.open("image.tmp", mode='wb')
+                    await f.write(await resp.read())
+                    await f.close()
+                    palette_color = await asyncio.get_running_loop().run_in_executor(None, functools.partial(fast_colorthief.get_palette, "image.tmp", color_count=2, quality=100))
```
Line is getting long, consider splitting it out to variables, ie loop, partial func, then assembling to run_in_executor.
Also, since you use it multiple times put the image name in a variable

Soitora replied:
:notlikethis:
How about we just split the line a little
```py
palette_color = await asyncio.get_running_loop().run_in_executor(
    None,
    functools.partial(
        fast_colorthief.get_palette,
        temp_img,
        color_count=2,
        quality=100
    )
)
```

doron-goldstein replied:
that works in the line limit but i really dislike it
also you can get the loop from the bot itself like so
```python
palette_partial = functools.partial(fast_colorthief.get_palette, "image.tmp", color_count=2, quality=100)
self.loop.run_in_executor(None, palette_partial)
```
can still break down the partial but thats up to you

doron-goldstein commented on bot.py:
```diff
@@ -25,26 +34,39 @@ async def format_embed(self, name, anime=False):
+                    await f.close()
+                    palette_color = await asyncio.get_running_loop().run_in_executor(None, functools.partial(fast_colorthief.get_palette, "image.tmp", color_count=2, quality=100))
+                    embed_color = "%02x%02x%02x" % palette_color[1]
+                    await aiofiles.os.remove("image.tmp")
```
Technically not really needed since we will always just override the existing file

Soitora replied:
Sure

Soitora committed:
Don't delete tempfile each run

Ignore the tempfile if we're not gonna be removing the file each run.

Includes sorting the .gitignore file so they are more in order.

Soitora committed:
Add AniList icon to footer

Soitora committed:
Revert "Don't italize genres"

This reverts commit a86bd78114d71f4819bc4e1535dc8e8e4aade4b8.

Soitora committed:
Don't need to import aiofiles.os anymore

Soitora committed:
Make image.tmp a variable

Soitora committed:
Remove the "Starting..." print

Soitora committed:
Handle AniList not returning 200 for cover

If it didn't return a 200 then it would crash due to embed_color being used before assignment.

Now it falls back to the AniList blue.

Image not loading is fine as Discord handles that naturally.

Soitora committed:
Split out the executor

Soitora committed:
Switch to async with context manager

No idea what I'm doing but it's running lmao

Soitora committed:
Change fallback color

Soitora committed:
Use kadal session instead of making a new one

Soitora committed:
Don't wordwrap in the middle of a word

Soitora committed:
Forgot the import...
